File: algorithms/555. Split Assembled Strings/solution.cc
class Solution {
public:
    string splitLoopedString(vector<string>& strs) {
        string result;
        vector<string> maxstrs;
        for ( int i = 0; i < strs.size(); ++i ) {
            string r(strs[i].rbegin(), strs[i].rend());
            maxstrs.push_back(max(strs[i], r));
        }
        for ( int i = 0; i < strs.size(); ++i ) {
            string mid;
            for ( int j = i + 1; j < strs.size(); ++j ) {
                mid += maxstrs[j];
            }
            for ( int j = 0; j < i; ++j ) {
                mid += maxstrs[j];
            }
            for ( int j = 0; j < strs[i].size(); ++j ) {
                result = max(result, strs[i].substr(j) + mid + strs[i].substr(0, j));
            }
            string r(strs[i].rbegin(), strs[i].rend());
            for ( int j = 0; j < strs[i].size(); ++j ) {
                result = max(result, r.substr(j) + mid + r.substr(0, j));
            }
        }
        return result;
    }
};

File: algorithms/555. Split Assembled Strings/solution.py
class Solution(object):
    def splitLoopedString(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        result = ""
        maxstrs = [max(s, s[::-1]) for s in strs]
        for i, word in enumerate(strs):
            mid = ''.join(maxstrs[i+1:] + maxstrs[:i])
            for j in xrange(len(word)):
                result = max(result, word[j:] + mid + word[:j])
            rword = word[::-1]
            for j in xrange(len(rword)):
                result = max(result, rword[j:] + mid + rword[:j])
        return result

# 49 / 49 test cases passed.
# Status: Accepted
# Runtime: 232 ms
